torch/distributions/half_cauchy.py

```diff
@@ -0,0 +1,55 @@
+import math
+from numbers import Number
+
+import torch
+from torch.distributions import constraints
+from torch.distributions.distribution import Distribution
+from torch.distributions import Cauchy
+from torch.distributions.utils import broadcast_all
+
+
+class HalfCauchy(Distribution):
+    r"""
+    Half-Cauchy distribution.
+
+    This is a continuous distribution with lower-bounded domain (`x > mu`).
+    See also the `Cauchy` distribution.
+
+    Example::
+
+        >>> m = HalfCauchy(torch.Tensor([0.0]), torch.Tensor([1.0]))
+        >>> m.sample()  # sample from a Cauchy distribution with loc=0 and scale=1
+        >>> m.sample()  # sample from HalfCauchy distributed with loc=0 and scale=1
+         0.3141
+        [torch.FloatTensor of size 1]
+
+    Args:
+        loc (float or Tensor or Variable): lower bound of the distribution.
+        scale (float or Tensor or Variable): half width at half maximum.
+    """
+
+    params = {'loc': constraints.real, 'scale': constraints.positive}
+    support = constraints.positive
```

Comment:
Replace this with a property below `__init__()`. See `Uniform` for example.
```py
@constraints.dependent_property
def support(self):
    return constraints.greater_than(self.loc)
```

Reply:
what does the decorator mean? how is this different than say `support = constraints.greater_than(self.loc)`

Reply:
Class variables cannot depend on self, hence the need to define a property. The fancy `@constraints.dependent_property` decorator is just like `@property` but adds a wrapper class so that the unbound property (e.g. `HalfCauchy.support`) is treated as a `constraints.dependent` but when bound to an instance (e.g. `HalfCauchy(0,1).support`) it is treated as a `constraints.greater_than(...)`. This pattern allows us to use cheaper static checks on most classes and use more expensive dynamic checks on classes with truly dependent properties.

If `HalfCauchy` is never used in practice with nonzero `loc`, you might consider dropping the `loc` parameter so that `.support` can stay static.